File: services/go/api/v1/auth/logout.go
package auth

import (
	"api/models"
	. "api/utils"
	. "api/v1/commons"
	"errors"

	"github.com/getsentry/sentry-go"
	"github.com/gofiber/fiber/v2"
	"gorm.io/gorm"
)

func Logout(c *fiber.Ctx) error {
	authToken := GetAuthToken(c)
	client := GetDB()

	var device models.Device

	if result := client.First(&device, "auth_token = ? ", authToken); result.Error != nil {
		if !errors.Is(result.Error, gorm.ErrRecordNotFound) {
			sentry.CaptureException(result.Error)
			return c.Status(500).JSON(Response{
				Success: false,
				Message: "There was an internal server error whilst logging you out!",
			})
		}
	}

	if result := client.Delete(&device); result.Error != nil {
		if !errors.Is(result.Error, gorm.ErrRecordNotFound) {
			sentry.CaptureException(result.Error)
			return c.Status(500).JSON(Response{
				Success: false,
				Message: "There was an internal server error whilst logging you out!",
			})
		}
	}

	_, err := RedisDel(authToken)

	if err != nil {
		sentry.CaptureException(err)
		return c.Status(500).JSON(Response{
			Success: false,
			Message: "There was an internal server error whilst deleting your token!",
		})
	}

	return c.JSON(Response{
		Success: true,
		Message: "Successfully deleted your session!",
	})
}

File: apps/api/src/handlers/auth/logout.ts
import { Id } from "@imperial/commons";
import { FastifyImp } from "../../types";
import { AuthSessions } from "../../utils/authSessions";

export const logout: FastifyImp = async (request, reply) => {
  const token = request.authentication_token;

  if (!token?.startsWith("imperial_auth_")) {
    return reply.status(400).send({
      success: false,
      error: {
        message: "You can not log out an API token",
      },
    });
  }

  await AuthSessions.deleteDeviceByAuthToken(token as Id<"imperial_auth">);

  reply.status(204).send();
};
